File: 02.09/baekjoon_29811_박동현.go
package main

import (
	"bufio"
	"container/heap"
	"os"
	"strconv"
)

// Heap
type Node struct {
	value, idx int
}
type HeapQ []Node

func (h HeapQ) Len() int {
	return len(h)
}
func (h HeapQ) Less(i, j int) bool {
	if h[i].value < h[j].value {
		return true
	} else if h[i].value == h[j].value {
		return h[i].idx < h[j].idx
	} else {
		return false
	}
}

func (h HeapQ) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *HeapQ) Push(x interface{}) {
	*h = append(*h, x.(Node))
}

func (h *HeapQ) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func makeEnumeratedHeap(idx int, lst []int, hq *HeapQ) {
	for i, v := range lst {
		heap.Push(hq, Node{v, i + idx})
	}
}

var (
	N     int
	arr   []int
	brr   []int
	arrHq = &HeapQ{}
	brrHq = &HeapQ{}
)

// Query
func U(x, y int) {
	if x <= N {
		heap.Push(arrHq, Node{y, x})
		arr[x-1] = y
	} else {
		heap.Push(brrHq, Node{y, x})
		brr[x-N-1] = y
	}
}

func L() {
	for (*arrHq)[0].value != arr[(*arrHq)[0].idx-1] {
		heap.Pop(arrHq)
	}
	for (*brrHq)[0].value != brr[(*brrHq)[0].idx-N-1] {
		heap.Pop(brrHq)
	}
	bw.WriteString(strconv.Itoa((*arrHq)[0].idx) + " " + strconv.Itoa((*brrHq)[0].idx) + "\n")
}

func main() {
	bs.Split(bufio.ScanWords)
	defer bw.Flush()
	N = inputInt()
	M := inputInt()
	// arr
	arr = make([]int, N)
	for i := range arr {
		arr[i] = inputInt()
	}
	heap.Init(arrHq)
	makeEnumeratedHeap(1, arr, arrHq)
	// brr
	brr = make([]int, M)
	for i := range brr {
		brr[i] = inputInt()
	}
	heap.Init(brrHq)
	makeEnumeratedHeap(N+1, brr, brrHq)

	// Query
	Q := inputInt()
	for i := 0; i < Q; i++ {
		cmd := input()
		if cmd == "U" {
			x, y := inputInt(), inputInt()
			U(x, y)
		} else {
			L()
		}
	}
}

// fastIO
var (
	bs = bufio.NewScanner(os.Stdin)
	bw = bufio.NewWriter(os.Stdout)
)

func input() string {
	bs.Scan()
	return bs.Text()
}

func inputInt() int {
	num, _ := strconv.Atoi(input())
	return num
}

File: 25/2/02.09/baekjoon_29811_박동현.py
import sys; input = sys.stdin.readline
from heapq import heappop, heappush


def make_enumerated_heap(lst,idx=1):
    res = []
    for k in lst:
        heappush(res, (k,idx))
        idx += 1
    return res

def U(x,y):
    x,y = map(int,(x,y))
    if x <= N:
        heappush(arr_hq, (y, x))
        arr[x-1]=y
    else:
        heappush(brr_hq, (y, x))
        brr[x-N-1]=y

def L(*args):
    while arr_hq[0][0] != arr[arr_hq[0][1]-1]:
        heappop(arr_hq)
    
    while brr_hq[0][0] != brr[brr_hq[0][1]-1-N]:
        heappop(brr_hq)
    print(arr_hq[0][1], brr_hq[0][1])

cmd = {
    "U": U,
    "L": L
}

N,M = map(int,input().split())
arr = [*map(int,input().split())]
brr = [*map(int,input().split())]

arr_hq = make_enumerated_heap(arr)
brr_hq = make_enumerated_heap(brr, N+1)

for _ in range(int(input())):
    c, *args = input().split()
    cmd[c](*args)